WordPress/gutenberg
# Theme.json: Remove custom prefixes from properties that did not land in 5.8

Related: https://github.com/WordPress/gutenberg/pull/31366#issuecomment-907075652, https://github.com/WordPress/gutenberg/pull/31366#issuecomment-908135526

## Description

This PR removes the `custom` prefix from properties within theme.json that do not require it. Those that remain have already landed in WP 5.8 and will be addressed in the future by adding a migration from `v1` to the future `v2` theme.json schema.

## How has this been tested?

- Unit tests
- Inspecting settings via the store in block and site editors
- Manually testing various combinations of prefixed and unprefixed flags
- Switching between themes using old and new flags

#### Testing instructions

1. Run the following tests:
```
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php
```
```
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-schema-v0-test.php
```
```
npm run test-unit-php /var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-schema-v1-test.php
```
	

2. Setup your theme.json to turn all the border and typography supports in the snippet below to `false`. 
    [Example gist here](https://gist.github.com/aaronrobertshaw/4934e8f308d1d747fef6ee7726440137)
```json
{
	"settings": {
		"border": {
			"color": false,
			"customRadius": false,
			"style": false,
			"width": false
		},
		"typography": {
			"fontStyle": false,
			"fontWeight": false,
			"textDecoration": false,
			"textTransform": false,
			"letterSpacing": false,
		},
	}
}
```
3. Create a new post and add a group block and navigation block to the post and save
4. Select the group block and ensure the border panel and its controls are not present
5. Confirm toggling the individual border supports on in your theme.json is reflected in the available controls when selecting the group block
6. Select the navigation block and ensure that none of the following controls show in the typography panel:
	* `fontStyle`
	* `fontWeight`
	* `textDecorations`
	* `textTransforms`
	* `letterSpacing`
7. Toggle each of the above properties on in your theme.json and confirm their related controls now display in the navigation block’s typography panel.
8. Switch to the site editor and ensure the correct controls also display in the relevant block panels there.
	* Note: Text decoration and text-transform have not been exposed within global styles to date. 
9. Change individual support flags back to their `custom` prefixed form e.g. `customFontStyle` or `customColor`. Then confirm the relevant controls still display correctly in both editors.
10. Test that `useSetting` will return the correct value when using the deprecated `custom` prefixed flag.
<details>
  <summary>Example patch to test `useSetting` with deprecated flags</summary>
  
```diff
diff --git a/packages/block-editor/src/hooks/border.js b/packages/block-editor/src/hooks/border.js
index 761c04200a..37a5213c27 100644
--- a/packages/block-editor/src/hooks/border.js
+++ b/packages/block-editor/src/hooks/border.js
@@ -23,17 +23,17 @@ export function BorderPanel( props ) {
 	const isSupported = hasBorderSupport( props.name );
 
 	const isColorSupported =
-		useSetting( 'border.color' ) && hasBorderSupport( props.name, 'color' );
+		useSetting( 'border.customColor' ) && hasBorderSupport( props.name, 'color' );
 
 	const isRadiusSupported =
-		useSetting( 'border.radius' ) &&
+		useSetting( 'border.customRadius' ) &&
 		hasBorderSupport( props.name, 'radius' );
 
 	const isStyleSupported =
-		useSetting( 'border.style' ) && hasBorderSupport( props.name, 'style' );
+		useSetting( 'border.customStyle' ) && hasBorderSupport( props.name, 'style' );
 
 	const isWidthSupported =
-		useSetting( 'border.width' ) && hasBorderSupport( props.name, 'width' );
+		useSetting( 'border.customWidth' ) && hasBorderSupport( props.name, 'width' );
 
 	if ( isDisabled || ! isSupported ) {
 		return null;
@@ -110,10 +110,10 @@ export function shouldSkipSerialization( blockType ) {
  */
 const useIsBorderDisabled = () => {
 	const configs = [
-		! useSetting( 'border.color' ),
-		! useSetting( 'border.radius' ),
-		! useSetting( 'border.style' ),
-		! useSetting( 'border.width' ),
+		! useSetting( 'border.customColor' ),
+		! useSetting( 'border.customRadius' ),
+		! useSetting( 'border.customStyle' ),
+		! useSetting( 'border.customWidth' ),
 	];
 
 	return configs.every( Boolean );

```
</details>

## Types of changes
Enhancement.
Breaking change ( 3rd parties working directly with theme.json or store shape will need updating )

## Checklist:
